### DISCUSSION TITLE
Measuring zero voltage and negative voltage

### DISCUSSION THREAD

**Discussion by Pandagineer** [bitcraze/discussions#332] [Q&A] [answered]
I am successfully measuring the voltage of an external 1.6 V battery between pins A3 and GND. Code is below for reference. Also, I have a flow deck attached.

My questions:

1. When I disconnect the external battery, the output sometimes goes to zero (roughly 10 mV) as expected, but then sometimes jumps to 3V. In practice, I will need the signal to stay at zero volts when I my sensor is reading nothing. (I don't mind averaging to filter noise, but if it jumps between zero and 3V, this would not work.) What is the best way to measure zero volts?

2. When I apply -1.6V (i.e. I reverse the external battery), the Console stops echoing, and the M4 LED turns off. Will doing this damage the CF? If negative voltage is bad, how low is ok (i.e. can I give it -1 mV? -10 mV?) What is happening here?

Thank you.

==============================

```
#include <string.h>
#include <stdint.h>
#include <stdbool.h>

#include "app.h"

#include "FreeRTOS.h"
#include "task.h"

#include "deck_constants.h"
#include "deck_analog.h"

float readMyAnalogPin() {
  return analogReadVoltage(DECK_GPIO_MISO);
}

#include "debug.h"

double y;

void appMain() {
  DEBUG_PRINT("Waiting for activation ...\n");

  while(1) {
    vTaskDelay(M2T(500));
    y = readMyAnalogPin();
    DEBUG_PRINT("%f", y);
    DEBUG_PRINT("\n");

  }
}
```

**Comment by Pandagineer** [comment, marked as answer]
Update on Question #1: I am now outputting all five A0-A4 channels. With no external battery attached, I am getting (roughly steady):  0.8 V, 0.8 V, 0.3 V, 0.8 V, and 0.5 V for the five channels.

Again, I would like these to be closer to zero. I will note that the sensor I eventually want to use will only output 10 mV.

**Reply by tobbeanton** [reply]
I think the issue you are having is that the pin is floating when there is no voltage connected. What you need to do is to pull this to a known voltage with a high value resistor. So if you want the ADC to read zero when battery is disconnected connect e.g. a 10k-10M ohm resistor to GND. Higher value resistor is better to minimize the effect it does to the voltage you really want to read.

10mV is pretty low. With 12bit ADC setting and 3.0V reference you will get 3.0V/2048 = 1.46mV per bit and the noise is several bits. Might be hard to measure well without amplification.

**Reply by Pandagineer** [reply]
Thank you. I'll try the resistor.
